pageserver/src/tenant.rs

```diff
@@ -1122,10 +1169,44 @@ impl Tenant {
         // FIXME original collect_timeline_files contained one more check:
         //    1. "Timeline has no ancestor and no layer files"
 
-        for (timeline_id, local_metadata) in sorted_timelines {
-            self.load_local_timeline(timeline_id, local_metadata, init_order, ctx)
+        // Process loadable timelines first
+        for (timeline_id, local_metadata) in scan.sorted_timelines_to_load {
+            if let Err(e) = self
+                .load_local_timeline(timeline_id, local_metadata, init_order, ctx, false)
                 .await
-                .with_context(|| format!("load local timeline {timeline_id}"))?;
+            {
+                match e {
+                    LoadLocalTimelineError::Load(source) => {
+                        return Err(anyhow::anyhow!(source)
+                            .context("Failed to load local timeline: {timeline_id}"))
+                    }
+                    LoadLocalTimelineError::ResumeDeletion(source) => {
+                        // Make sure resumed deletion wont fail loading for entire tenant.
+                        error!("Failed to resume timeline deletion: {source:#}")
+                    }
+                }
+            }
+        }
+
+        // Resume deletion ones with deleted_mark
+        for (timeline_id, local_metadata) in scan.timelines_to_resume_deletion {
+            if let Err(e) = self
+                .load_local_timeline(timeline_id, local_metadata, init_order, ctx, true)
+                .await
+            {
+                match e {
+                    LoadLocalTimelineError::Load(source) => {
+                        // We tried to load deleted timeline, this is a bug.
+                        return Err(anyhow::anyhow!(source).context(
+                            "This is a bug. We tried to load deleted timeline which is wrong and loading failed. Timeline: {timeline_id}"
+                        ));
+                    }
+                    LoadLocalTimelineError::ResumeDeletion(source) => {
+                        // Make sure resumed deletion wont fail loading for entire tenant.
+                        error!("Failed to resume timeline deletion: {source:#}")
+                    }
+                }
+            }
```

Comment:
I don't like the re-use of `load_local_timeline`, and these two loops here.
Reads quite hacky and unintuitive.

I guess the reason why you did it this way is that we can late-discover the need to delete a timeline, i.e., we only learn that we need to delete it after we've fetched the remote index, because the local marker is missing.

Why not lift the index part download out of `load_local_timelien` into this function here? That way, you have cleanly disjoint sets. One to sort&load, the other to resume deletion.

Should also allow getting rid of that pesky `found_delete_mark: bool` naked bool argument.

Reply:
> Reads quite hacky and unintuitive.

To me it feels rather clear. At this point we have two maps and we just execute the intention.

> I guess the reason why you did it this way is that we can late-discover the need to delete a timeline, i.e., we only learn that we need to delete it after we've fetched the remote index, because the local marker is missing.

Yep

> Why not lift the index part download out of load_local_timelien into this function here? That way, you have cleanly disjoint sets. One to sort&load, the other to resume deletion.

I think if I lift the index part download to this function there wont be anything left out of load_local_timeline. Aside from call to fetch the ancestor and `timeline_init_and_sync`. Currently it looks like a rather clear separation of concerns. We have the scan logic that gathers what it has in terms of local fs and passes the control over to loading which then applies remote parts of the equation. Maybe renaming of `load_local_timeline` can help.

I can try to shuffle things around a bit, maybe better in a follow up and we can see whether it is better than what we have there.

Reply:
I can also change the bool to be an enum so it is more clear whats going on

Reply:
Created https://github.com/neondatabase/neon/issues/4648

Reply:
Thank you!